### DISCUSSION TITLE
Access to data pixel directly

### DISCUSSION THREAD

**Discussion by zsdennis** [libvips/libvips#3894] [General]
Hi, 

i need know if possible access to data directly line by line from VipImage.

VipsSource * in;
in = vips_source_new_from_file( inPath );
unsigned char *buffer;
buffer = (unsigned char*)malloc(xs);
vips_source_read( in , buffer , xs);

Now i use a VipsSource but in this case i load VipsSource from file, but i need work directly from an existing VipsImage.

And i would read a buffer, and not pixel by pixle with VIPS_IMAGE_ADDR().

Thank for help

**Comment by jcupitt** [comment]
Hi @zsdennis, there are several ways to do this, it depends what you want to do.

## Force the whole image into memory

This can be very slow, and obviously can use a lot of memory, but it's very simple. Use `vips_image_wio_input()`, then `VIPS_IMAGE_ADDR()` to get a pointer to a pixel.

You can use `vips_image_inplace()` to force an image into memory for read-write.

## Threaded line by line read

You can get a libvips worker to execute a function you define for every "chunk" of pixels in the image, have a look at `vips_sink()`.

There's also `vips_sink_tile()` to generate in tiles, and `vips_sink_disc()` to execute your function sequentially in a write-behind manner from a worker.

## Sequential line by line read

You can make a `VipsRegion`, then ask for those pixels to be generated with `vips_region_prepare()`. You won't get any threading, but it's also very simple.

**Comment by zsdennis** [comment]
Hi @jcupitt ,

thank for fast reply.

Ok, but i need increase speed.

With vips_region_prapare() or other method, how get entire scanline, without get pixel x pixel with VIPS_IMAGE_ADDR() ?

Similar to this : 

vips_source_read( in , buffer , xs);

but with VipsImage and not source.

Thank

**Reply by jcupitt** [reply]
Yes, they all give full pointer access to pixels, so you can scan around using regular pointer arithmetic. Use `VIPS_REGION_LSKIP()` or `VIPS_IMAGE_LSKIP()` to get the number of bytes to add to move down a line. For `VipsRegion` access, you can read all the pixels within the rect `region->valid`.

You can only read pixels. You need to use `_inplace()` to get write access too, but that will cause pipelines to stop and restart, so be careful.

**Comment by zsdennis** [comment]
hi @jcupitt ,

i try this, but execution is extremly low.

```C
VipsRegion *region = vips_region_new(_in);

for (int y = 0; y < ys; y++){      
        
        VipsRect r = { .left = 0, .top = y, .width = wD, .height = 1 };
        vips_region_prepare(region, &r)
        memcpy(ibuf, region->data,xs);

}
```

i tryed to make region large same as image dimansion, and move in a region pointer, but i have an warning:

```
(process:9380): VIPS-WARNING **: 15:06:08.470: out of memory --- size == 66MB
```

however not increse speed.

**Reply by jcupitt** [reply]
It should be quick. Make a complete test program I can try.

**Reply by jcupitt** [reply]
... with a test image too, or notes on how to make a test image.

**Comment by jcupitt** [comment]
... I made you a tiny demo program:

```C
/* compile with:
 *
 * gcc -g -Wall try350.c `pkg-config vips --cflags --libs`
 */

#include <stdio.h>
#include <vips/vips.h>

int
main(int argc, char **argv)   
{
    VipsImage *image;

    if (VIPS_INIT(argv[0]))
        vips_error_exit(NULL);

    if (!(image = vips_image_new_from_file(argv[1], "access", VIPS_ACCESS_SEQUENTIAL, NULL)))
        vips_error_exit(NULL);

    VipsRegion *region = vips_region_new(image);
    int sum = 0;
    for (int y = 0; y < image->Ysize; y++) {
        if (vips_region_prepare(region, &(VipsRect) { 0, y, image->Xsize, 1 }))
            vips_error_exit(NULL);

        unsigned char *p = VIPS_REGION_ADDR(region, 0, y);
        int size = VIPS_REGION_SIZEOF_LINE(region);
        for (int x = 0; x < size; x++) 
            sum += p[x];
    }

    printf("sum = %d\n", sum);

    g_object_unref(region);
    g_object_unref(image);

    return 0;
}
```

That will only work for 8-bit images, and I've not tried to handle errors or int overflow correctly. With a 10k x 10k pixel RGB JPEG I see:

```
$ /usr/bin/time -f %M:%e ./a.out ~/pics/wtc.jpg
sum = 990128018
152832:0.71
```

So about 700ms to decompress and scan a 300mb image, with a peak memory use of 150mb.

That won't use any threading. You can speed it up further with `vips_sink_disc()`, but that's a little more complicated.

**Comment by zsdennis** [comment]
Hi @jcupitt ,

sorry for delay.

in my case, with :

VipsRect r = {0, y,  xs,  1 };
vips_region_prepare(region, &r);*/
memcpy(ibuf, region->data,xs);

time is 23 sec, since.

so the images are load in ACCESS_RANDOM, and is one band.

thank

**Reply by jcupitt** [reply]
You need to post a complete program I can try, or post a link to a sample image that I can try my demo on.

If I can't reproduce your slow performance, I can't help :(
